File: medium/1833_maximum_ice_cream_bars/solution.cpp
#include <vector>

class Solution {
public:
  int maxIceCream(std::vector<int> &costs, int coins) {
    std::vector<int> costsCounter(
        *std::max_element(costs.begin(), costs.end()) + 1, 0);
    for (const int &cost : costs)
      ++costsCounter[cost];

    int iceCreamCount = 0;

    for (int cost = 1; cost < costsCounter.size(); ++cost) {
      int count = costsCounter[cost];
      if (!count)
        continue;

      long iceCreamsCost = static_cast<long>(cost) * count;
      if (coins >= iceCreamsCost) {
        coins -= iceCreamsCost;
        iceCreamCount += count;
      } else {
        while (coins >= cost) {
          coins -= cost;
          ++iceCreamCount;
        }
      }
    }

    return iceCreamCount;
  }
};

File: medium/1833_maximum_ice_cream_bars/Solution.java
class Solution {
  public int maxIceCream(int[] costs, int coins) {
    int maxCost = costs[0];
    for (final int cost : costs) {
      maxCost = Math.max(maxCost, cost);
    }

    final int[] costsCounter = new int[maxCost + 1];
    for (final int cost : costs) {
      ++costsCounter[cost];
    }

    int coinsCopy = coins;
    int iceCreamCount = 0;

    for (int cost = 1; cost < costsCounter.length; ++cost) {
      final int count = costsCounter[cost];
      if (count == 0) {
        continue;
      }

      final int iceCreamsCost = cost * count;
      if (coinsCopy >= iceCreamsCost) {
        coinsCopy -= iceCreamsCost;
        iceCreamCount += count;
      } else {
        while (coinsCopy >= cost) {
          coinsCopy -= cost;
          ++iceCreamCount;
        }
      }
    }

    return iceCreamCount;
  }
}
